### PR TITLE
Aws provider

### PR DESCRIPTION
**Pull request by owlishDeveloper**
I created a buggie https://bugzilla.mozilla.org/show_bug.cgi?id=1575094

Also, I'm not sure what provider metadata is useful for debugging, so I just put in there some stuff that's similar to what is done for gcp. Some metadata that gcp has is about networking, and aws iid don't have that (except for private ip which I'm uncomfortable exposing). We would need to make another aws api call to get that..... I decided to leave that out for now - we can always add that if we need to

### CODE REVIEW

File: provider/google/google.go [hunk 1]
```diff
@@ -47,7 +47,9 @@ func (p *GoogleProvider) ConfigureRun(state *run.State) error {
 		return fmt.Errorf("Could not create worker manager client: %v", err)
 	}
 
-	err = provider.RegisterWorker(state, wm, userData.WorkerPoolID, userData.ProviderID, userData.WorkerGroup, workerID, "token", proofToken)
+	workerIdentityProofMap := map[string]interface{}{"token": interface{}(proofToken)}
+
+		err = provider.RegisterWorker(state, wm, userData.WorkerPoolID, userData.ProviderID, userData.WorkerGroup, workerID, workerIdentityProofMap)
```
Comment (djmitche) [inline provider/google/google.go]: I think this needs a `gofmt` run.  I'm not sure why the `lint` job didn't catch this??

File: provider/awsprovider/awsprovider.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	proto *protocol.Protocol
+}
+
+func (p *AWSProvider) ConfigureRun(state *run.State) error {
+	userData, err := p.metadataService.queryUserData()
+	if err != nil {
+		return fmt.Errorf("Could not query user data: %v", err)
+	}
+
+	instanceIdentityDocument_string, instanceIdentityDocument_json, err := p.metadataService.queryInstanceIdentityDocument()
```
Comment (djmitche) [inline provider/awsprovider/awsprovider.go]: I'm surprised lint didn't catch these mixed-format identifiers, either :)

**Comment by owlishDeveloper** [comment]
Yesterday I was trying to understand what `providerData` is for, and found this comment https://github.com/taskcluster/taskcluster-worker-runner/blob/eaae08f4b11ac13777453941941ad280a44e757f/run/state.go#L29-L32 <--- it led me to believe that this info is needed exclusively to print in logs for debugging purposes. I haven't realized docker-worker actually needs it. I would probably add a comment that this is not a debugging info but a public api (as in, something that another service is expecting, and if it doesn't get it, there might be Consequences)

**Comment by owlishDeveloper** [comment]
I looked at docker-worker - it seems to be getting the metadat itself as well?
https://github.com/taskcluster/docker-worker/blob/0ffd8bdb44be082e5cbf8c3c4b3ec4eea0976628/src/lib/host/aws.js#L72-L111

File: go.sum [hunk 3]
```diff
@@ -31,6 +31,7 @@ github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94/go.mod h1:AZpEONHx3
 github.com/stretchr/objx v0.1.0/go.mod h1:HFkY916IF+rwdDfMAkV7OtwuqBVzrE8GR6GFx+wExME=
 github.com/stretchr/testify v1.3.0 h1:TivCn/peBQ7UY8ooIcPgZFpTNSz0Q2U6UrFlUfqbe0Q=
 github.com/stretchr/testify v1.3.0/go.mod h1:M5WIy9Dh21IEIfnGCwXGc5bZfKNJtfHm1UVUgZn+9EI=
+github.com/stretchr/testify v1.4.0 h1:2E4SXV/wtOkTonXsotYi4li6zVWxYlZuYNCXe9XRJyk=
```
Comment (owlishDeveloper) [inline go.sum]: why are there two versions of the same thing? Dustin, what do I do?

**Comment by djmitche** [comment]
> I looked at docker-worker - it seems to be getting the metadata itself as well?

`src/lib/host/aws.js` is one of a few "host" implementations in docker-worker, and one of the others is `taskcluster-workerunner.js`, which is what we use when running under tc-worker-runner.  Once everything's using that host implementation, we'll delete `aws.js`.  But, for the moment, it serves as a good reference for what data docker-worker needs (I used it when writing `worker/dockerworker/dockerworker.go`).

File: go.sum (hunk 3 above)
Comment (djmitche) [reply go.sum]: My general rule with `go.sum` is "don't worry about it", much like `yarn.lock`.  I'm not sure why it does this, but it seems happy about it..

File: run/state.go [hunk 4]
```diff
@@ -27,9 +27,11 @@ type State struct {
...
 
 	// metadata from the provider (useful to display to the user for
-	// debugging).  Workers should not *require* any data to exist
-	// in this map, and where possible should just pass it along as-is
-	// in worker config as helpful debugging metadata for the user.
+	// debugging).  This is also a public API for docker-worker,
+	// which doesn't, strictly speaking, require these fields,
+	// but may fall onto undesirable defaults if these are not provided
+	// A bit more info on that here
+	// https://github.com/taskcluster/taskcluster-worker-runner/pull/30#pullrequestreview-277378260
```
Comment (djmitche) [inline run/state.go]: Looks good.  @milescrabill you may want to include here the fields that generic-worker requires, too.

**Comment by owlishDeveloper** [comment]
🐑!

**Comment by djmitche** [comment]
Do you need a release made?  If not, I'll wait for my changes and @milescrabill to land and then make one.

**Pull request merged**